### DISCUSSION TITLE
Inter-class constraint between variables from list of random classes

### DISCUSSION THREAD

**Discussion by ShreyanJabade** [fvutils/pyvsc#76] [General]
Hello,
We are using PyVSC module for constrained randomization in Python. There is a parent class (`Parent()`) which has a list of random classes as its instance variables (`c1` and `c2`). It is needed to specify a constraint between two variables from two different class instances from the list (say `c1[0].a` and `c2[0].x`). Doing this however gives `ExprLiteralModel' object has no attribute 'fm'`. The code snippet used is attached below:
```
import vsc 
from vsc import *
@vsc.randobj
class Parent:
    def __init__(self):
        self.id = 0
        self.c1 = vsc.rand_list_t(vsc.attr(Child1()))
        for i in range(10):    
            self.c1.append(vsc.attr(Child1()))

        self.c2 = vsc.rand_list_t(vsc.attr(Child2()))
        for i in range(10):
            self.c2.append(vsc.attr(Child2()))

    @vsc.constraint
    def ab_c(self):
        self.c1[0].a < 6
        self.c1[0].a == self.c2[0].x

@vsc.randobj
class Child1():
    def __init__(self):
        self.a = vsc.rand_uint8_t()
        self.b = vsc.rand_uint8_t()
    
    @vsc.constraint
    def test_c(self):
        self.a < 3

@vsc.randobj
class Child2():
    def __init__(self):
        self.x = vsc.rand_uint8_t()
        self.y = vsc.rand_uint8_t()
    
    @vsc.constraint
    def test_c(self):
        self.x < 6

inst=Parent()
inst.randomize()
print(inst.i1[0].a)
print(inst.i2[1].a)
```

I request you to look into this aspect.
Thank you.

**Comment by mballance** [comment]
Hello @ShreyanJabade,
    Thank you for the bug report and, especially, the testcase! After a quick look, I don't see anything that looks incorrect with the code and an internal error like you're seeing (object has not attribute...) certainly indicates a bug. I'll investigate when I have time later today.

Best Regards,
Matthew

**Comment by mballance** [comment]
Thanks for raising this issue, @ShreyanJabade. I investigated and found an issue with using literal indexes into arrays of classes. Strangely enough, using an equivalent constraint inside a foreach constraint worked.
I've corrected the issue and your testcase now runs properly. The latest source on GitHub contains the fix, as does the 0.4.0 release.

Best Regards,
Matthew

**Reply by ShreyanJabade** [reply]
Hello, 
Thank you for your swift consideration. The aspect mentioned above is working as expected.
I extended the above testcase to a more general case as attached in snippet below. Here, each of the child classes (`c1` and `c2`) now have a list of classes (`self.c1[i].a` is now a list of random `Field` classes). Like a multi-level hierarchy. Doing this gives `'FieldArrayModel' object has no attribute 'type_f'`. Also, `foreach` constraints can be present (e.g. in `parent_c` constraint ) and there can be multiple levels in general. Is there is a restriction on how many levels deep we can go?
I request you to look into this aspect.
Thank you

Code snippet:

```
import vsc 
from vsc import *
@vsc.randobj
class Parent:
    def __init__(self):
        self.id = 0
        self.c1 = vsc.rand_list_t(vsc.attr(Child1()))
        for i in range(10):    
            self.c1.append(vsc.attr(Child1()))      

        self.c2 = vsc.rand_list_t(vsc.attr(Child2()))
        for i in range(10):
            self.c2.append(vsc.attr(Child2()))

    @vsc.constraint
    def parent_c(self):
        self.c1[0].a[1].value == self.c2[0].x[1].value       # Multi-level 

        with vsc.foreach(self.c1, idx=True) as i:
            self.c1[i].a[0].value == self.c2[i].x[0].value


@vsc.randobj
class Field():
    def __init__(self, name, def_value):
        self.name = name
        self.value = vsc.rand_uint8_t(def_value)

@vsc.randobj
class Child1:
    def __init__(self):
        self.a = vsc.rand_list_t(vsc.attr(Field('a', 10)))
        for i in range(5):    
            self.a.append(vsc.attr(Field('a', 10)))

        self.b = vsc.rand_list_t(vsc.attr(Field('b', 10)))
        for i in range(5):    
            self.b.append(vsc.attr(Field('b', 10)))
    
    @vsc.constraint
    def test_c(self):
        self.a[0].value < self.a[1].value

@vsc.randobj
class Child2:
    def __init__(self):
        self.x = vsc.rand_list_t(vsc.attr(Field('x', 10)))
        for i in range(5):    
            self.x.append(vsc.attr(Field('x', 10)))

        self.y = vsc.rand_list_t(vsc.attr(Field('y', 10)))
        for i in range(5):    
            self.y.append(vsc.attr(Field('y', 10)))
    
    @vsc.constraint
    def test_c(self):
        self.x[0].value < self.x[1].value

inst=Parent()
inst.randomize()
print(inst.c1[0].a[0].value)
print(inst.c2[0].x[0].value)

```

**Comment by mballance** [comment]
Thank you for the new testcase, @ShreyanJabade. I've investigated, and have been able to reproduce the original problem. However, I'll need to investigate further to solve the core problem. While PyVSC has no intentional limit on the number of nesting levels with class arrays, it appears that currently it is limited to a single level (class containing an array of classes). Again, this is not an intentional limit, and I will correct this issue. Unfortunately, it will take me more time that I have to spend on the issue this evening.
Just wanted to give you an update that I am able to reproduce the issue, and am making progress.

Best Regards,
Matthew

**Comment by mballance** [comment]
Hi @ShreyanJabade,
    I've made a fix in the 0.4.1 release that resolves the issues with your testcase above. Please try this version and let me know how it works for you. I will plan to do a bit of extended testing as well.

Best Regards,
Matthew

**Reply by ShreyanJabade** [reply]
Hello,
Thank you for considering this aspect. On running the same testcase used in comment above (attached herewith for ready reference), it gives "[pybtor] [boolector] boolector_ult: argument 'e0' belongs to different Boolector instance" with the latest release.

Version Info: 
![image](https://user-images.githubusercontent.com/58798198/122590225-e0bef600-d07e-11eb-8dd5-026c474f9543.png)

The screenshot of the error at my end:

![image](https://user-images.githubusercontent.com/58798198/122589944-80c84f80-d07e-11eb-9987-d3cd2374bdf2.png)

Code snippet used (same as above):
```
import vsc 
from vsc import *
@vsc.randobj
class Parent:
    def __init__(self):
        self.id = 0
        self.c1 = vsc.rand_list_t(vsc.attr(Child1()))
        for i in range(10):    
            self.c1.append(vsc.attr(Child1()))      

        self.c2 = vsc.rand_list_t(vsc.attr(Child2()))
        for i in range(10):
            self.c2.append(vsc.attr(Child2()))

    @vsc.constraint
    def parent_c(self):
        self.c1[0].a[1].value == self.c2[0].x[1].value       # Multi-level 

        with vsc.foreach(self.c1, idx=True) as i:
            self.c1[i].a[0].value == self.c2[i].x[0].value


@vsc.randobj
class Field():
    def __init__(self, name, def_value):
        self.name = name
        self.value = vsc.rand_uint8_t(def_value)

@vsc.randobj
class Child1:
    def __init__(self):
        self.a = vsc.rand_list_t(vsc.attr(Field('a', 10)))
        for i in range(5):    
            self.a.append(vsc.attr(Field('a', 10)))

        self.b = vsc.rand_list_t(vsc.attr(Field('b', 10)))
        for i in range(5):    
            self.b.append(vsc.attr(Field('b', 10)))
    
    @vsc.constraint
    def test_c(self):
        self.a[0].value < self.a[1].value

@vsc.randobj
class Child2:
    def __init__(self):
        self.x = vsc.rand_list_t(vsc.attr(Field('x', 10)))
        for i in range(5):    
            self.x.append(vsc.attr(Field('x', 10)))

        self.y = vsc.rand_list_t(vsc.attr(Field('y', 10)))
        for i in range(5):    
            self.y.append(vsc.attr(Field('y', 10)))
    
    @vsc.constraint
    def test_c(self):
        self.x[0].value < self.x[1].value

inst=Parent()
inst.randomize()
print(inst.c1[0].a[0].value)
print(inst.c2[0].x[0].value)
```

**Comment by mballance** [comment]
Apologies, @ShreyanJabade. I thought I was running your exact testcase. I will double-check to see if I inadvertently made a modification that masked an issue.

**Reply by mballance** [reply]
Okay, @ShreyanJabade, I've understood the issue and have a correction in the latest release (0.4.2). The testcase above now runs correctly for me. Please confirm whether it also runs correctly for you.

Best Regards,
Matthew
